File: tests/fixtures/selenium/java/LoginTest.java
package com.example.tests;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import java.time.Duration;
import org.testng.annotations.Test;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.AfterMethod;
import static org.testng.Assert.*;

public class LoginTest {
    WebDriver driver;

    @BeforeMethod
    public void setUp() {
        driver = new ChromeDriver();
        driver.manage().window().maximize();
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
    }

    @Test
    public void testSuccessfulLogin() {
        driver.get("https://example.com/login");

        WebElement usernameField = driver.findElement(By.id("username"));
        usernameField.clear();
        usernameField.sendKeys("testuser");

        WebElement passwordField = driver.findElement(By.id("password"));
        passwordField.clear();
        passwordField.sendKeys("password123");

        WebElement loginButton = driver.findElement(By.cssSelector(".login-btn"));
        loginButton.click();

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("dashboard")));

        WebElement welcomeMsg = driver.findElement(By.xpath("//h1[contains(text(), 'Welcome')]"));
        assertTrue(welcomeMsg.isDisplayed(), "Welcome message should be visible");
        assertEquals(driver.getTitle(), "Dashboard - MyApp");
    }

    @Test
    public void testFailedLogin() {
        driver.get("https://example.com/login");

        driver.findElement(By.id("username")).sendKeys("wronguser");
        driver.findElement(By.id("password")).sendKeys("wrongpass");
        driver.findElement(By.cssSelector(".login-btn")).click();

        Thread.sleep(2000);

        WebElement errorMsg = driver.findElement(By.className("error-message"));
        assertTrue(errorMsg.isDisplayed());
        assertTrue(errorMsg.getText().contains("Invalid credentials"));
    }

    @AfterMethod
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}

File: tests/fixtures/selenium/js/login.test.js
const { Builder, By, until, Key } = require('selenium-webdriver');
const assert = require('assert');

describe('Login Page', function () {
  let driver;

  before(async function () {
    driver = await new Builder().forBrowser('chrome').build();
  });

  after(async function () {
    await driver.quit();
  });

  it('should login successfully', async function () {
    await driver.get('https://example.com/login');

    const username = await driver.findElement(By.id('username'));
    await username.clear();
    await username.sendKeys('testuser');

    const password = await driver.findElement(By.id('password'));
    await password.clear();
    await password.sendKeys('password123');

    await driver.findElement(By.css('.login-btn')).click();

    await driver.wait(until.elementLocated(By.id('dashboard')), 10000);

    const title = await driver.getTitle();
    assert.strictEqual(title, 'Dashboard - MyApp');

    const welcome = await driver.findElement(By.xpath("//h1[contains(text(), 'Welcome')]"));
    const isDisplayed = await welcome.isDisplayed();
    assert.ok(isDisplayed);
  });

  it('should show error for bad credentials', async function () {
    await driver.get('https://example.com/login');

    await driver.findElement(By.id('username')).sendKeys('wrong');
    await driver.findElement(By.id('password')).sendKeys('wrong');
    await driver.findElement(By.css('.login-btn')).click();

    await driver.sleep(2000);

    const error = await driver.findElement(By.className('error-message'));
    const text = await error.getText();
    assert.ok(text.includes('Invalid credentials'));
  });

  it('should navigate to registration', async function () {
    await driver.get('https://example.com/login');

    await driver.findElement(By.linkText('Create Account')).click();

    await driver.wait(until.urlContains('/register'), 5000);

    const heading = await driver.findElement(By.css('h1'));
    const headingText = await heading.getText();
    assert.strictEqual(headingText, 'Register');
  });
});
